catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/Poller.java

```diff
@@ -0,0 +1,270 @@
+   *   <li>In order of completion time, the new value from each {@link Callable#call()} is loaded
+   *       into the cache, overwriting the old value. If {@link Callable#call()} timed out, {@link
+   *       #handleTimeout(Object)} is called for special timeout handling. If {@link
+   *       Callable#call()} threw a {@link RuntimeException}, {@link #handleException(Object,
+   *       RuntimeException)} is called for special exception handling.
+   * </ol>
+   *
+   * @throws IllegalArgumentException if {@code timeout} is less than {@value MINIMUM_TIMEOUT}
+   * @throws NullPointerException if {@code timeoutTimeUnit} or {@code current} is {@code null}
+   * @throws IllegalStateException if there is a threading error
+   * @throws IllegalStateException if {@link #pollThreadPool} or {@link
+   *     #pollTimeoutWatcherThreadPool} {@link * ScheduledExecutorService#isShutdown()}
+   */
+  /*final*/ synchronized void pollCurrent(
+      final long timeout, final TimeUnit timeoutTimeUnit, final Map<K, Callable<V>> current) {
+    if (timeout < MINIMUM_TIMEOUT) {
+      throw new IllegalArgumentException(
+          "timeout argument may not be less than " + MINIMUM_TIMEOUT);
+    }
+    notNull(timeoutTimeUnit);
+    notNull(current);
+    if (pollThreadPool.isShutdown() || pollTimeoutWatcherThreadPool.isShutdown()) {
+      throw new IllegalStateException(
+          "Unable to poll because pollThreadPool or pollTimeoutWatcherThreadPool is shutdown");
+    }
+
+    removeNoncurrentKeysFromTheCache(current.keySet());
+
+    final int currentCount = current.size();
+
+    // Loop through each of the current
+    LOGGER.trace("Starting a process to get the value for each of the {} current", currentCount);
+    final CompletionService<Runnable> completionService =
+        new ExecutorCompletionService<>(pollTimeoutWatcherThreadPool);
+    current.forEach(
+        (key, loader) -> {
+          try {
+            // Start the process to load a new value
+            final Future<V> future = pollThreadPool.submit(loader);
+
+            // Submit a separate process to wait for the load to finish or timeout
+            completionService.submit(
+                () -> getCommitRunnable(key, future, timeout, timeoutTimeUnit));
+          } catch (RejectedExecutionException e) {
+            throw new IllegalStateException("Unable to poll " + key, e);
+          }
+        });
+
+    // In order of loader process completion, commit the new value to the cache or handles the
+    // timeout or RuntimeException
+    for (int i = 0; i < currentCount; i++) {
+      final Runnable commitRunnable;
+      try {
+        commitRunnable = completionService.take().get();
+      } catch (CancellationException | ExecutionException e) {
+        throw new IllegalStateException(
+            "Exception while waiting for polling to complete or time out", e);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
```

Comment:
See https://github.com/codice/ddf/pull/3858#discussion_r248797923

Reply:
Interruption are only introduced when something else is interrupting. This is usually when we shutdown. We shouldn't be throwing another exception like this. There are 3 ways to handle interruptions:
1- declare that you throw it and let your caller handle the fact that you were interrupted.
2- catch it, remember you were interrupted, finish what you were doing, and at the end of it rethrow it or propagate the interruption by interrupting the current thread. This would typically be done outside of your work (or the loop) in a finally clause to make sure that no matter how you exit what you were doing, you reset that interruption state (or throw back the interrupted exception) properly
3- You stop what you were doing and either throw the interruption out for your caller to handle or re-interrupt the thread.

If you do some work and do not intend to finish it when an interruption is detected, you should typically throw back that exception or let your caller know you interrupted what you were doing and not return successfully or throw some other exception as done here.